File: app/src/main/java/lt/vilnius/tvarkau/AppModule.java
package lt.vilnius.tvarkau;

import android.app.Application;

import com.squareup.leakcanary.RefWatcher;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import lt.vilnius.tvarkau.dagger.module.IoScheduler;
import lt.vilnius.tvarkau.dagger.module.UiScheduler;
import lt.vilnius.tvarkau.fragments.presenters.ConnectivityProvider;
import lt.vilnius.tvarkau.fragments.presenters.ConnectivityProviderImpl;
import rx.Scheduler;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

@Module
public class AppModule {

    private TvarkauApplication application;

    AppModule(TvarkauApplication application) {
        this.application = application;
    }

    @Provides
    @Singleton
    Application providesApplication() {
        return application;
    }

    @Provides
    @Singleton
    @IoScheduler
    Scheduler provideIoScheduler() {
        return Schedulers.io();
    }

    @Provides
    @Singleton
    @UiScheduler
    Scheduler provideUiScheduler() {
        return AndroidSchedulers.mainThread();
    }

    @Provides
    @Singleton
    RefWatcher providesRefWatcher() {
        return application.getRefWatcher();
    }

    @Provides
    @Singleton
    ConnectivityProvider provideConnectivityProvider() {
        return new ConnectivityProviderImpl(application);
    }
}

File: app/src/main/java/lt/vilnius/tvarkau/dagger/module/AppModule.kt
package lt.vilnius.tvarkau.dagger.module

import com.squareup.leakcanary.RefWatcher
import dagger.Module
import dagger.Provides
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import lt.vilnius.tvarkau.TvarkauApplication
import lt.vilnius.tvarkau.dagger.DbScheduler
import lt.vilnius.tvarkau.dagger.IoScheduler
import lt.vilnius.tvarkau.dagger.UiScheduler
import lt.vilnius.tvarkau.fragments.presenters.ConnectivityProvider
import lt.vilnius.tvarkau.fragments.presenters.ConnectivityProviderImpl
import javax.inject.Named
import javax.inject.Singleton

@Module
class AppModule {

    @Provides
    @Singleton
    @IoScheduler
    fun provideIoScheduler(): Scheduler {
        return Schedulers.io()
    }

    @Provides
    @Singleton
    @UiScheduler
    fun provideUiScheduler(): Scheduler {
        return AndroidSchedulers.mainThread()
    }

    @Provides
    @Singleton
    fun providesRefWatcher(
        application: TvarkauApplication
    ): RefWatcher {
        return application.refWatcher
    }

    @Provides
    @Singleton
    @DbScheduler
    fun provideDbScheduler(): Scheduler {
        return Schedulers.single()
    }

    @Provides
    @Singleton
    fun provideConnectivityProvider(
        application: TvarkauApplication
    ): ConnectivityProvider {
        return ConnectivityProviderImpl(application)
    }

    @Provides
    @Named("all_reports")
    fun provideAllReportTypes(): String {
        return ""
    }
}
